src/Dataset.ts

```diff
@@ -0,0 +1,218 @@
+// A Dataset manages the production and manipulation of *tiles*.
+
+import { Tile, Rectangle, QuadTile } from './tile';
+import {
+  extent, range, min, max, bisectLeft,
+} from 'd3-array';
+import Zoom from './interaction';
+import * as Comlink from 'comlink';
+
+//@ts-ignore
+import TileWorker from './tileworker.worker.js?worker&inline';
+import { APICall } from './types';
+import Scatterplot from './deepscatter';
+type Key = string;
+
+export abstract class Dataset<T extends Tile> {
+  abstract root_tile : T;
+//  public mutations : 
+  public max_ix : number = -1;
+  protected plot : Scatterplot;
+  protected _tileworkers: TileWorker[] = [];
+  abstract ready : Promise<void>;
+  constructor(plot : Scatterplot) {
+    this.plot = plot;
+  }
+  static from_quadfeather(url : string, prefs: APICall, plot: Scatterplot) : QuadtileSet {
+    if (url.match(/(\/[0-9]+){3}/)) {
+      throw new Error('Quadfeathers must be loaded from a base URL.');
+    }
+    return new QuadtileSet(url, prefs, plot);
+  }
+
+  get extent() : Rectangle {
+    return {
+      x: [-1e16, 1e16],
+      y: [-1e16, 1e16],
+    }
+  }
+
+  download_most_needed_tiles(bbox : Rectangle, max_ix: number, queue_length : number) {
```

Comment:
If this method isn't implemented, shouldn't it be an abstract method?

Reply:
OK, I see how to do this in ts now. Yes, you're right.